#  Module 13 – Missing Values



## 1. isnull()

`isnull()` is a Pandas method used to identify missing or null values in a DataFrame or Series.

It returns `True` when a value is missing and `False` when a value is present.

### Real-World Example

A student dataset may have missing marks for a student who has not yet completed an exam.

### Business Example

A customer dataset may contain missing values for phone numbers, email addresses, or locations.

### AI/ML Example

Before training a Machine Learning model, it is important to identify missing values because many models require complete input data.

In [1]:
import pandas as pd

data = {
    "Name": ["Sai", "Rahul", "Priya"],
    "Age": [23, None, 20],
    "Marks": [85, 78, None]
}

df = pd.DataFrame(data)

print(df.isnull())

    Name    Age  Marks
0  False  False  False
1  False   True  False
2  False  False   True


### Explanation

The `isnull()` method checks every value in the DataFrame.

`True` means the value is missing, while `False` means the value is available.

In this example, the Age value for Rahul and the Marks value for Priya are missing.

### Best Practices

Use `isnull()` during initial data inspection to identify missing values.

### Common Mistakes

Do not assume that `False` means the value is zero. It only means that the value is not missing.

## 2. notnull()

`notnull()` is a Pandas method used to identify values that are not missing.

It returns `True` when a value is present and `False` when a value is missing.

### Real-World Example

A student dataset can be checked to find which students have marks available.

### Business Example

A company can identify customer records where important information such as email or phone number is available.

### AI/ML Example

`notnull()` can be used to select complete records or identify available values during data preprocessing.

In [2]:
print(df.notnull())

   Name    Age  Marks
0  True   True   True
1  True  False   True
2  True   True  False


### Explanation

The `notnull()` method returns `True` for values that are present.

The missing Age value for Rahul and missing Marks value for Priya are shown as `False`.

### Best Practices

Use `notnull()` when you need to identify or select values that are available.

### Common Mistakes

Confusing `notnull()` with `isnull()`. They return opposite Boolean results.

## 3. dropna()

`dropna()` is a Pandas method used to remove rows or columns that contain missing values.

It is useful when missing records cannot be reliably filled or when incomplete records are not required.

### Real-World Example

A student record with important missing information may be removed from the dataset.

### Business Example

A company may remove records that are missing critical information required for a particular analysis.

### AI/ML Example

During preprocessing, rows with missing values may be removed when only a small number of records are affected and removing them will not significantly reduce the dataset.

In [3]:
cleaned_df = df.dropna()

print(cleaned_df)

  Name   Age  Marks
0  Sai  23.0   85.0


### Explanation

The `dropna()` method removes rows containing one or more missing values.

Only the row for Sai remains because it has values in all three columns.

### Best Practices

Use `dropna()` when the number of missing records is small and removing them will not affect the analysis significantly.

### Common Mistakes

Removing too many rows can cause useful data to be lost.

## 4. fillna()

`fillna()` is a Pandas method used to replace missing values with a specified value.

It is useful when we want to keep the records instead of removing them.

### Real-World Example

Missing student marks can be replaced with a suitable value such as the class average.

### Business Example

Missing sales values can sometimes be filled using a calculated value such as an average or a known default.

### AI/ML Example

Missing numerical features can be filled using appropriate statistical values before training a Machine Learning model.

In [4]:
filled_df = df.copy()

filled_df["Age"] = filled_df["Age"].fillna(
    filled_df["Age"].mean()
)

filled_df["Marks"] = filled_df["Marks"].fillna(
    filled_df["Marks"].mean()
)

print(filled_df)

    Name   Age  Marks
0    Sai  23.0   85.0
1  Rahul  21.5   78.0
2  Priya  20.0   81.5


### Explanation

The missing Age value is replaced with the average Age.

The missing Marks value is replaced with the average Marks.

This allows us to keep all three student records.

### Best Practices

Choose the fill value based on the meaning and distribution of the data.

### Common Mistakes

Filling missing values with an unsuitable value can introduce bias into the dataset.

## 5. interpolate()

`interpolate()` is a Pandas method used to estimate missing values based on nearby values.

It is especially useful for ordered or continuous data such as time-series measurements.

### Real-World Example

If a temperature reading is missing between two recorded temperatures, interpolation can estimate the missing value.

### Business Example

A company may use interpolation to estimate missing daily sales values between known dates.

### AI/ML Example

Interpolation can be useful when preparing time-series data where missing values occur between known observations.

In [5]:
data = {
    "Day": [1, 2, 3, 4, 5],
    "Sales": [100, None, 140, None, 180]
}

sales_df = pd.DataFrame(data)

sales_df["Sales"] = sales_df["Sales"].interpolate()

print(sales_df)

   Day  Sales
0    1  100.0
1    2  120.0
2    3  140.0
3    4  160.0
4    5  180.0


### Explanation

The `interpolate()` method estimates missing values using the values around them.

The missing Sales value on Day 2 is estimated as 120, and the missing value on Day 4 is estimated as 160.

### Best Practices

Use interpolation when the data has a meaningful order or continuous pattern.

### Common Mistakes

Using interpolation for data where the order of values has no meaningful relationship.

## 6. Missing Value Handling Strategies

Missing Value Handling means deciding how to manage missing data before analysis or Machine Learning.

Common strategies include removing missing records, filling missing values with statistical values, using forward or backward filling, and using interpolation when appropriate.

The correct strategy depends on the amount of missing data, the type of variable, and the purpose of the analysis.

### Real-World Example

If only a few student records have missing marks, they may be removed or the missing values may be filled using a suitable method.

### Business Example

A company may replace missing numerical sales values with an appropriate statistical estimate, while missing categorical values may be replaced with the most suitable category or an "Unknown" label.

### AI/ML Example

Missing values must be handled carefully before model training because inappropriate treatment can reduce model quality or introduce bias.

In [6]:
data = {
    "Name": ["Sai", "Rahul", "Priya", "Amit"],
    "Age": [23, None, 20, 21],
    "Marks": [85, 78, None, 88]
}

df = pd.DataFrame(data)

print("Missing values:")
print(df.isnull().sum())

df["Age"] = df["Age"].fillna(df["Age"].mean())
df["Marks"] = df["Marks"].fillna(df["Marks"].mean())

print("\nAfter handling missing values:")
print(df)

Missing values:
Name     0
Age      1
Marks    1
dtype: int64

After handling missing values:
    Name        Age      Marks
0    Sai  23.000000  85.000000
1  Rahul  21.333333  78.000000
2  Priya  20.000000  83.666667
3   Amit  21.000000  88.000000


### Explanation

First, `isnull().sum()` counts the missing values in each column.

The missing Age and Marks values are then replaced with the mean of their respective columns.

Different datasets may require different strategies. We should understand the data before choosing how to handle missing values.

### Best Practices

Choose a missing value strategy based on the data type, amount of missing data, and purpose of the analysis.

### Common Mistakes

Using the same missing value strategy for every column without considering the meaning of the data.

## Advantages

- Helps identify missing data clearly.
- Provides different methods for handling missing values.
- Helps keep datasets usable for analysis.
- Supports better data preparation for Machine Learning.
- Can reduce problems caused by incomplete datasets.

## Limitations

- Removing missing rows can lead to loss of useful data.
- Filling values with averages can change the original data distribution.
- Incorrect interpolation can produce unrealistic values.
- Different columns may require different handling strategies.
- Poor handling of missing values can introduce bias.

## Interview Questions and Answers

### 1. What are missing values?

Missing values are data points where information is not available or has not been recorded.

### 2. What does `isnull()` do?

`isnull()` identifies missing values and returns `True` for missing values.

### 3. What does `notnull()` do?

`notnull()` identifies values that are present and returns `True` for non-missing values.

### 4. What is `dropna()` used for?

`dropna()` removes rows or columns containing missing values.

### 5. What is `fillna()` used for?

`fillna()` replaces missing values with a specified value.

### 6. What is `interpolate()` used for?

`interpolate()` estimates missing values using nearby values, especially in ordered data.

### 7. What is the difference between `dropna()` and `fillna()`?

`dropna()` removes missing records, while `fillna()` replaces missing values.

### 8. When should missing rows be removed?

They can be removed when the number of missing records is small and removing them will not significantly affect the analysis.

### 9. Why is missing value handling important in AI/ML?

Many Machine Learning algorithms require complete input data, and poor handling of missing values can affect model performance.

### 10. What factors should be considered before handling missing values?

We should consider the amount of missing data, the data type, the importance of the column, and the purpose of the analysis.

## Self Reflection

I learned how to identify and handle missing values using Pandas.

I practiced `isnull()`, `notnull()`, `dropna()`, `fillna()`, and `interpolate()`.

I also understood that different datasets require different missing value handling strategies.

Overall, I gained a better understanding of how missing data can be managed before Data Analysis and Machine Learning.